toolchain/check/call.cpp

```diff
@@ -313,6 +314,10 @@ auto PerformCall(Context& context, SemIR::LocId loc_id, SemIR::InstId callee_id,
     // Preserve the `self` argument from the original callee.
     CARBON_CHECK(!callee_function.self_id.has_value());
     callee_function.self_id = self_id;
+
+    // Special handling for C++ member overloaded operators.
+    AdjustSelfAndArgsForCppMemberOverloadedOperator(
+        context, callee_function.function_id, callee_function.self_id, arg_ids);
```

Comment:
@zygoloid Does this seem like the right way to handle this to you? Calling back into the C++ layers right after `PerformCppOverloadResolution` just feels odd to me.

@bricknerb This function name is also pretty verbose; it feels like the current function name would be better as a comment. Had you considered something like:

```suggestion
  if (self_id.has_value()) {
    callee_function.self_id = self_id;
  } else if (IsCppOperatorMethod(context, callee_function.function_id)) {
    // Adjust `self` and args for C++ overloaded operator methods.
    callee_function.self_id = arg_ids.consume_front();
  }
```

BTW, to be sure, it's specifically methods, not just that it's a member (including non-instance members, such as `friend auto operator>(T lhs, T rhs) -> bool`)?

Reply:
I think converting the call into a member call is fine for now. We'll need to revisit this when we start generating `OpWith` impls anyway -- though at that point I think what we'll want to do is generate a method-shaped thunk for a non-method-shaped operator function, which is sort of the opposite of what's happening here. But I see no need to do the work for that now :)

Reply:
That said: I do wonder whether we could do all the work here in `PerformCppOverloadResolution`, and change its return type to be something like a `std::pair<InstId, CalleeFunction>`.

Reply:
Thanks!
I've made one commit to change it to use `IsCppOperatorMethod()` and another commit to move everything into `PerformCppOverloadResolution()`, so you can see what makes more sense.
I think hiding the C++ specific logic from the non `cpp/` files makes sense.
Note that I now also have to return the args from `PerformCppOverloadResolution()`, since they could be modified. Alternatively I can pass args by reference to mutate them.

Regarding methods vs. member function, I think I don't understand some terminology.
IIUC, `friend auto operator>(T lhs, T rhs) -> bool` is not a member function.
Can you clarify how do we use the term `method` compared to `member function` in the Carbon code base?

Reply:
> Regarding methods vs. member function, I think I don't understand some terminology. IIUC, `friend auto operator>(T lhs, T rhs) -> bool` is not a member function. Can you clarify how do we use the term `method` compared to `member function` in the Carbon code base?

Yes, `friend` functions are not really members. I probably shouldn't have called it that, I was really just wanting you to be more precise about "method". Note though that C++ is getting [static overloaded operators](https://en.cppreference.com/w/cpp/language/operators.html). So there's a distinction between a member operator and method operator that's significant.

"method" is shorthand for "instance member function". I believe this is the same way you'd call it in C++; e.g., this code checks against `CXXMethodDecl`.

C++ has static member functions, Carbon often calls those "class functions".

Carbon terminology is defined [in class docs](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/classes.md#member-functions).

Reply:
Thanks for the reference!
Definitely helpful given it's slightly different from C++ standard terminology.
FWIW, `CXXMethodDecl` could be static.
https://github.com/llvm/llvm-project/blob/cae73be8a6c2d39e5a827fd31848676d52062a94/clang/include/clang/AST/DeclCXX.h#L2125

Reply:
FWIW, the merge with the latest changes (#6108 and #6117 specifically)  was non trivial.

* Since we decided to move the logic of setting `self_id` inside `PerformCppOverloadResolution()`, I moved the `CARBON_KIND_SWITCH()` on the overload result callee to `PerformCppOverloadResolution()` and made the return value struct contain an optional `CalleeFunction`.
I'm a bit concerned about extra copies here, but didn't find a nice way to avoid it in the scope of this PR. Feel free to suggest further changes which I will follow up.

* I had to adjust the num params value when calling `ImportCppFunctionDecl()` for operator methods. We could also do it inside `ImportCppFunctionDecl()` but it feels like here we have more context since we do the arg to self conversion here. There might be alternative solutions which I'm not aware of.

Reply:
Sent out #6122 for `PerformCallToCppOverloadFunction()`.